x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/BlobCacheMetrics.java

```diff
@@ -56,4 +104,55 @@ public LongCounter getEvictedCountNonZeroFrequency() {
     public LongHistogram getCacheMissLoadTimes() {
         return cacheMissLoadTimes;
     }
+
+    /**
+     * Record the various cache population metrics after a chunk is copied to the cache
+     *
+     * @param totalBytesCopied The total number of bytes copied
+     * @param totalCopyTimeNanos The time taken to copy the bytes in nanoseconds
+     * @param index The index being loaded
+     * @param shardId The ID of the shard being loaded
+     * @param cachePopulationReason The reason for the cache being populated
+     * @param cachePopulationSource The source from which the data is being loaded
+     */
+    public void recordCachePopulationMetrics(
+        int totalBytesCopied,
+        long totalCopyTimeNanos,
+        String index,
+        int shardId,
+        CachePopulationReason cachePopulationReason,
+        CachePopulationSource cachePopulationSource
+    ) {
+        Map<String, Object> metricAttributes = Map.of(
+            INDEX_ATTRIBUTE_KEY,
+            index,
+            SHARD_ID_ATTRIBUTE_KEY,
+            shardId,
+            CACHE_POPULATION_REASON_ATTRIBUTE_KEY,
+            cachePopulationReason.name(),
+            CACHE_POPULATION_SOURCE_ATTRIBUTE_KEY,
+            cachePopulationSource.name()
+        );
+        assert totalBytesCopied > 0 : "We shouldn't be recording zero-sized copies";
+        cachePopulationBytes.incrementBy(totalBytesCopied, metricAttributes);
+
+        // This is almost certainly paranoid, but if we had a very fast/small copy with a very coarse nanosecond timer it might happen?
+        if (totalCopyTimeNanos > 0) {
+            cachePopulateThroughput.record(toMebibytesPerSecond(totalBytesCopied, totalCopyTimeNanos), metricAttributes);
+            cachePopulationTime.incrementBy(TimeUnit.NANOSECONDS.toMillis(totalCopyTimeNanos), metricAttributes);
+        }
+    }
+
+    /**
+     * Calculate throughput as MiB/second
+     *
+     * @param totalBytes The total number of bytes transferred
+     * @param totalNanoseconds The time to transfer in nanoseconds
+     * @return The throughput as MiB/second
+     */
+    private double toMebibytesPerSecond(int totalBytes, long totalNanoseconds) {
+        double totalSeconds = totalNanoseconds / 1_000_000_000.0;
+        double totalMegabytes = totalBytes / 1_048_576.0;
+        return totalMegabytes / totalSeconds;
```

Comment:
Nit: Since we are using double here, it's unlikely to have overflow/underflow issue here. But I wonder whether it would still be better to compute this as `bytes/nanos * const` where `const  = 1e9 / 2**20`. If we assume 16MB copied bytes which is in the magnitude of 1e7 bytes and 1ms copy time which is in the magnitude of 1e6 nanos, the computation of `bytes/nanos * const` most of the times not involving too large or small numbers, which feels better to me.

Reply:
Addressed in b17f316